segmentstore/storage/src/main/java/io/pravega/segmentstore/storage/chunklayer/ChunkedSegmentStorage.java

```diff
@@ -653,6 +669,58 @@ private void close(String message, AutoCloseable toClose) {
         }
     }
 
+    /**
+     * Adds block index entries.
+     */
+    void addBlockIndexEntries(MetadataTransaction txn, String segmentName, long startOffset, long endOffset, List<ChunkNameOffsetPair> readIndexEntries) {
+        val blockSize = config.getIndexBlockSize();
+        val startBlock = startOffset / blockSize;
+        val endBlock = endOffset / blockSize;
+        TreeMap<Long, String> map = new TreeMap<>();
+        readIndexEntries.stream().forEach( pair -> map.put(pair.getOffset(), pair.getChunkName()));
+        for (long block = startBlock; block <= endBlock; block++) {
+            val blockStartOffset = block * blockSize;
+            val indexEntry = map.floorEntry(blockStartOffset);
+            if (null != indexEntry) {
+                val blockEntry = ReadIndexBlockMetadata.builder()
+                        .name(NameUtils.getSegmentReadIndexBlockName(segmentName, blockStartOffset))
+                        .startOffset(indexEntry.getKey())
+                        .chunkName(indexEntry.getValue())
+                        .status(StatusFlags.ACTIVE)
+                        .build();
+                txn.create(blockEntry);
+                log.debug("{} adding new block index entry segment={}, entry={}.", logPrefix, segmentName, blockEntry);
+            }
+        }
+    }
+
+    /**
+     * Adds block index entries for given chunk.
+     */
+    void addBlockIndexEntries(MetadataTransaction txn, String segmentName, String chunkName, long startOffset, long endOffset) {
+        val blockSize = config.getIndexBlockSize();
```

Comment:
You should validate your arguments before making any changes. Both here and in the other 2 methods.